api/views/phones.py

```diff
@@ -622,39 +626,246 @@ def _get_phone_objects(inbound_to):
+    match_type: Optional[Literal["short", "full", "last"]] = None
+    # The prefix portion of the text message, empty string if no prefix
+    prefix: str = ""
+    # The detected short code or full number, or None if none detected
+    detected: Optional[str] = None
+    # The matching numbers, as e.164 strings, empty if None
+    numbers: list[str] = field(default_factory=list)
+
+
+@dataclass
+class MatchData(MatchByPrefix):
+    """Details of expanding a MatchByPrefix with InboundContacts."""
+
+    # The matching InboundContacts
+    contacts: list[InboundContact] = field(default_factory=list)
+
+
+def _match_senders(relay_number: RelayNumber, text: str) -> MatchData:
+    """Find the likely InboundContact(s) specified in the SMS text."""
+    multi_replies_flag, _ = get_waffle_flag_model().objects.get_or_create(
+        name="multi_replies",
+        defaults={
+            "note": (
+                "MPP-2252: Use prefix on SMS text to specify the recipient,"
+                " rather than default of last contact."
+            )
+        },
+    )
+
+    if (
+        multi_replies_flag.is_active_for_user(relay_number.user)
+        or multi_replies_flag.everyone
+    ):
+        # Load all the previous contacts, collect possible countries
+        contacts = InboundContact.objects.filter(relay_number=relay_number).all()
+        contacts_by_number: dict[str, InboundContact] = {}
+        for contact in contacts:
+            pn = phonenumbers.parse(contact.inbound_number)
+            e164 = phonenumbers.format_number(pn, phonenumbers.PhoneNumberFormat.E164)
+            if e164 not in contacts_by_number:
+                contacts_by_number[e164] = contact
+
+        match = _match_by_prefix(text, set(contacts_by_number.keys()))
+        if match.match_type in {"short", "full"}:
+            return MatchData(
+                contacts=[contacts_by_number[num] for num in match.numbers],
+                **asdict(match),
+            )
+
+    last_text_sender = get_last_text_sender(relay_number)
+    if last_text_sender:
+        return MatchData(match_type="last", contacts=[last_text_sender])
+    else:
+        return MatchData(match_type=None)
+
+
+_SMS_SHORT_PREFIX_RE = re.compile(
+    r"""
+^               # Start of string
+\s*             # One or more spaces
+\d{4}           # 4 digits
+\s*             # Optional whitespace
+[-:/\\\]|]?     # At most one separator, sync with SMS_SEPARATORS below
+\s*             # Trailing whitespace
+""",
+    re.VERBOSE | re.ASCII,
+)
+_SMS_SEPARATORS = set("-:/\\]|")  # Sync with SMS_SHORT_PREFIX_RE above
+
+
+def _match_by_prefix(text: str, candidates: set[str]) -> MatchByPrefix:
+    """
+    Look for a prefix in a text message, and return a likely phone number match.
+
+    Arguments:
+    * A SMS text message
+    * A set of phone numbers in E.164 format
+    """
+    # Gather potential countries, needed by PhoneNumberMatcher
+    countries = set()
```

Comment:
suggestion (non-blocking): re-name this variable to `region_codes` to be more clear what's in here?